ulog.cpp

```diff
@@ -0,0 +1,350 @@
+            _logging_data_process(ulog_data);
+        }
+        break;
+    }
+    }
+
+    return buffer->len;
+}
+
+bool ULog::_logging_seq(uint16_t seq, bool *drop)
+{
+    /* Check for message drops */
+    if (_expected_seq != seq) {
+        *drop = true;
+
+        /*
+         * Handle case were Flight Stack did not received a ACK and sent the
+         * same message again.
+         */
+        if (_expected_seq - 1 == seq)
+            return false;
+
+        _expected_seq = seq;
+    }
+
+    _expected_seq++;
+    return true;
+}
+
+void ULog::_logging_data_process(mavlink_logging_data_t *msg)
+{
+    bool drops = false;
+
+    if (!_logging_seq(msg->sequence, &drops))
+        return;
+
+    /* Waiting for ULog header? */
+    if (_waiting_header) {
+        const uint8_t magic[] = ULOG_MAGIC;
+
+        if (memcmp(magic, msg->data, sizeof(magic))) {
+            log_error("Invalid ULog Magic number");
+            return;
+        }
+        if (msg->length < ULOG_HEADER_SIZE) {
+            /* This should never happen */
+            log_error("ULog header is not completed");
+            return;
+        }
+
+        const ssize_t r = write(_file, msg->data, ULOG_HEADER_SIZE);
+        if (r != ULOG_HEADER_SIZE) {
+            log_error_errno(errno, "Unable to write to ULog file: (%m)");
+            return;
+        }
+
+        memmove(msg->data, &msg->data[ULOG_HEADER_SIZE], msg->length);
+        msg->length -= ULOG_HEADER_SIZE;
+        _waiting_header = false;
+    }
+
+    if (drops) {
+        _logging_flush();
+
+        _buffer_len = 0;
+        _waiting_first_msg_offset = true;
+    }
+
+    /*
+     * Do not cause a buffer overflow, it should only happens if a ULog message
+     * don't fit in _msg_buffer
+     */
+    if (_buffer_len + msg->length >= sizeof(_buffer)) {
+        log_warning("Buffer full, dropping everything on buffer");
+
+        _buffer_len = 0;
+        _waiting_first_msg_offset = true;
+    }
+
+    uint8_t begin = 0;
+
+    if (_waiting_first_msg_offset) {
+        if (msg->first_message_offset == NO_FIRST_MSG_OFFSET) {
+            /* no useful information in this message */
+            return;
+        }
+
+        _waiting_first_msg_offset = false;
+        begin = msg->first_message_offset;
+    }
+
+    if (!msg->length)
+        return;
+
+    memcpy(&_buffer[_buffer_len], &msg->data[begin], msg->length - begin);
+    _buffer_len += (msg->length - begin);
+    _logging_flush();
+}
+
+void ULog::_logging_flush()
+{
+    while (_buffer_len >= sizeof(struct ulog_msg_header)) {
+        struct ulog_msg_header *header = (struct ulog_msg_header *)_buffer;
+        const uint16_t full_msg_size = header->msg_size + sizeof(struct ulog_msg_header);
+
+        if (full_msg_size > _buffer_len) {
+            break;
+        }
+
+        const ssize_t r = write(_file, _buffer, full_msg_size);
+        if (r == -EAGAIN)
+            return;
+        if (r != full_msg_size) {
```

Comment:
we potentially wrote part of the message. We should still recalculate _buffer_len below or we will corrupt the log when we enter this function again on next write.